src/hotspot/share/runtime/os.cpp

```diff
@@ -1764,13 +1764,104 @@ char* os::attempt_reserve_memory_at(char* addr, size_t bytes, bool executable) {
   char* result = pd_attempt_reserve_memory_at(addr, bytes, executable);
   if (result != nullptr) {
     MemTracker::record_virtual_memory_reserve((address)result, bytes, CALLER_PC);
+    log_debug(os)("Reserved memory at " INTPTR_FORMAT " for " SIZE_FORMAT " bytes.", p2i(addr), bytes);
   } else {
     log_debug(os)("Attempt to reserve memory at " INTPTR_FORMAT " for "
                  SIZE_FORMAT " bytes failed, errno %d", p2i(addr), bytes, get_last_error());
   }
   return result;
 }
 
+char* os::get_lowest_attach_address() {
+  return (char*)os::vm_allocation_granularity();
+}
+
+char* os::get_highest_attach_address() {
+  return (char*)(
+#ifdef _LP64
+  (128 * 1024 * G)
+#else
+  SIZE_MAX
+#endif
+  - os::vm_page_size());
+}
```

Comment:
I am not sure what "attach" means in this sense. If it's the usable address range, wouldn't it need to be OS-specific?

Reply:
Yes, highest and lowest usable address. You are right in that its OS specific. I'll move these to the respective OS files.